File: 110-balanced-binary-tree/110-balanced-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        def helper(curr: Optional[TreeNode]) -> int:
            if not curr:
                return 0
            
            left = helper(curr.left)
            right = helper(curr.right)
            
            if abs(right - left) > 1:
                return inf
            else:
                return max(left, right) + 1
            
        ans = helper(root)
        return True if ans != inf else False

File: 110-balanced-binary-tree/110-balanced-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isBalanced(TreeNode root) {
        int ans = this.helper(root);
        return ans > -1;
    }
    
    private int helper(TreeNode curr) {
        if(curr == null){
            return 0;
        }
        
        int left = this.helper(curr.left);
        int right = this.helper(curr.right);
        
        if(left == -1 || right == -1 || Math.abs(right - left) > 1){
            return -1;
        } else {
            return Math.max(left, right) + 1;
        }
    }
}